File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/AlertListContainer.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.util.fromHex
import com.mbta.tid.mbta_app.android.util.modifiers.haloContainer
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.AlertCardSpec
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.RouteType

@Composable
fun AlertListContainer(
    highPriority: List<@Composable (modifier: Modifier) -> Unit> = listOf(),
    middleContent: (@Composable (modifier: Modifier) -> Unit)? = null,
    lowPriority: List<@Composable (modifier: Modifier) -> Unit> = listOf(),
) {
    val highPriorityCount = highPriority.size
    val lowPriorityCount = lowPriority.size
    val outerCornerRadius = 8.dp
    val internalCornerRadius = 4.dp
    Column(
        modifier = Modifier.haloContainer(2.dp, backgroundColor = Color.Transparent),
        verticalArrangement = Arrangement.spacedBy(1.dp),
    ) {
        highPriority.forEachIndexed { index, content ->
            val topRadius = if (index == 0) outerCornerRadius else internalCornerRadius
            val bottomRadius =
                if (index == highPriorityCount && (middleContent == null && lowPriorityCount == 0))
                    outerCornerRadius
                else internalCornerRadius
            content(
                Modifier.background(
                    colorResource(R.color.fill3),
                    RoundedCornerShape(
                        topStart = topRadius,
                        topEnd = topRadius,
                        bottomStart = bottomRadius,
                        bottomEnd = bottomRadius,
                    ),
                )
            )
        }

        val middleContentTopRadius =
            if (highPriorityCount == 0) outerCornerRadius else internalCornerRadius
        val middleContentBottomRadius =
            if (lowPriorityCount == 0) outerCornerRadius else internalCornerRadius

        middleContent?.invoke(
            Modifier.background(
                colorResource(R.color.fill3),
                RoundedCornerShape(
                    topStart = middleContentTopRadius,
                    topEnd = middleContentTopRadius,
                    bottomStart = middleContentBottomRadius,
                    bottomEnd = middleContentBottomRadius,
                ),
            )
        )

        lowPriority.forEachIndexed { index, content ->
            val topRadius =
                if (index == 0 && highPriorityCount == 0 && middleContent == null) outerCornerRadius
                else internalCornerRadius
            val bottomRadius =
                if (index == highPriorityCount) outerCornerRadius else internalCornerRadius
            content(
                Modifier.background(
                    colorResource(R.color.fill2),
                    RoundedCornerShape(
                        topStart = topRadius,
                        topEnd = topRadius,
                        bottomStart = bottomRadius,
                        bottomEnd = bottomRadius,
                    ),
                )
            )
        }
    }
}

@Composable
@Preview
fun AlertListContainerPreview() {
    Column(
        verticalArrangement = Arrangement.spacedBy(8.dp),
        modifier = Modifier.background(Color.Blue),
    ) {
        AlertListContainer(
            highPriority =
                listOf({ modifier ->
                    AlertCard(
                        ObjectCollectionBuilder.Single.alert({
                            effect = Alert.Effect.ServiceChange
                        }),
                        null,
                        AlertCardSpec.Basic,
                        routeAccents =
                            TripRouteAccents(
                                color = Color.fromHex("80276C"),
                                textColor = Color.fromHex("FFFFFF"),
                                type = RouteType.COMMUTER_RAIL,
                            ),
                        modifier = modifier,
                        onViewDetails = {},
                    )
                }),
            middleContent = { modifier -> NotAccessibleCard(modifier) },
            lowPriority =
                listOf({ modifier ->
                    AlertCard(
                        ObjectCollectionBuilder.Single.alert({
                            effect = Alert.Effect.ElevatorClosure
                            header =
                                "Ruggles Elevator 848 (Lobby to lower busway side platform) unavailable due to maintenance"
                        }),
                        null,
                        AlertCardSpec.Elevator,
                        routeAccents =
                            TripRouteAccents(
                                color = Color.fromHex("ED8B00"),
                                textColor = Color.fromHex("FFFFFF"),
                                type = RouteType.HEAVY_RAIL,
                            ),
                        modifier = modifier,
                        onViewDetails = {},
                    )
                }),
        )

        AlertListContainer(
            highPriority =
                listOf({ modifier ->
                    AlertCard(
                        ObjectCollectionBuilder.Single.alert({
                            effect = Alert.Effect.ServiceChange
                        }),
                        null,
                        AlertCardSpec.Basic,
                        routeAccents =
                            TripRouteAccents(
                                color = Color.fromHex("80276C"),
                                textColor = Color.fromHex("FFFFFF"),
                                type = RouteType.COMMUTER_RAIL,
                            ),
                        modifier = modifier,
                        onViewDetails = {},
                    )
                })
        )

        AlertListContainer(
            lowPriority =
                listOf({ modifier ->
                    AlertCard(
                        ObjectCollectionBuilder.Single.alert({
                            effect = Alert.Effect.ServiceChange
                        }),
                        null,
                        AlertCardSpec.Basic,
                        routeAccents =
                            TripRouteAccents(
                                color = Color.fromHex("80276C"),
                                textColor = Color.fromHex("FFFFFF"),
                                type = RouteType.COMMUTER_RAIL,
                            ),
                        modifier = modifier,
                        onViewDetails = {},
                    )
                })
        )
    }
}

File: iosApp/iosApp/ComponentViews/AlertListContainer.swift
//
//  AlertListContainer.swift
//  iosApp
//
//  Created by Kayla Brady on 3/24/26.
//  Copyright © 2026 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct AlertListContainer: View {
    @ObserveInjection var inject
    let displayAlerts: DisplayAlerts
    let showNotAccessibleCard: Bool
    let now: EasternTimeInstant
    let isAllServiceDisrupted: Bool
    let routeIdMatcher: Matcher<Route.Id>
    let stopIdMatcher: Matcher<NSString>
    let directionIdMatcher: Matcher<KotlinInt>
    let tripId: String?
    let routeAccents: TripRouteAccents
    let onRowTap: (String, AlertCardSpec) -> Void

    var body: some View {
        VStack(spacing: 1) {
            let highPriorityCount = displayAlerts.highPriority.count
            let lowPriorityCount = displayAlerts.lowPriority.count
            let outerCornerRadius: CGFloat = 8
            let internalCornerRadius: CGFloat = 4

            ForEach(Array(displayAlerts.highPriority.enumerated()), id: \.element.id) { index, displayAlert in
                let topRadius = index == 0 ? outerCornerRadius : internalCornerRadius
                let bottomRadius = index == highPriorityCount - 1 && !showNotAccessibleCard && lowPriorityCount ==
                    0 ? outerCornerRadius : internalCornerRadius

                alertCard(displayAlert)
                    .background(Color.fill3)
                    .withUnevenRoundedBorder(topRadius: topRadius, bottomRadius: bottomRadius, color: Color.clear)
            }

            if showNotAccessibleCard {
                let topRadius = highPriorityCount == 0 ? outerCornerRadius : internalCornerRadius
                let bottomRadius = lowPriorityCount == 0 ? outerCornerRadius : internalCornerRadius

                NotAccessibleCard()
                    .background(Color.fill3)
                    .withUnevenRoundedBorder(topRadius: topRadius, bottomRadius: bottomRadius, color: Color.clear)
            }
            ForEach(Array(displayAlerts.lowPriority.enumerated()), id: \.element.id) { index, displayAlert in
                let topRadius = index == 0 && highPriorityCount == 0 && !showNotAccessibleCard ? outerCornerRadius
                    : internalCornerRadius
                let bottomRadius = index == lowPriorityCount - 1 ? outerCornerRadius : internalCornerRadius

                alertCard(displayAlert)
                    .background(Color.fill2)
                    .withUnevenRoundedBorder(topRadius: topRadius, bottomRadius: bottomRadius, color: Color.clear)
            }
        }
        .padding(2)
        .background(Color.halo)
        .clipShape(RoundedRectangle(cornerRadius: 10))
        .enableInjection()
    }

    @ViewBuilder
    func alertCard(_ displayAlert: DisplayAlert) -> some View {
        let alert = displayAlert.alert
        let tripIdMatcher: Matcher<NSString> = if let tripId { MatcherData(value: tripId as NSString) }
        else { MatcherWildcard() }
        let _ = debugPrint("AAAAA", alert, routeIdMatcher, stopIdMatcher, directionIdMatcher, tripIdMatcher)
        let summaryEntity = alert.summary(
            routeId: routeIdMatcher,
            stopId: stopIdMatcher,
            directionId: directionIdMatcher,
            tripId: tripIdMatcher,
        )
        let spec = displayAlert.cardSpec(now: now, isAllServiceDisrupted: isAllServiceDisrupted, tripId: tripId)

        AlertCard(
            alert: alert,
            alertSummaryEntity: summaryEntity,
            spec: spec,
            routeAccents: routeAccents,
            onViewDetails: { onRowTap(alert.id, spec) }
        )
    }
}

#Preview {
    let serviceChange = ObjectCollectionBuilder.Single.shared.alert { alert in
        alert.effect = .serviceChange
    }

    let elevator = ObjectCollectionBuilder.Single.shared.alert { alert in
        alert.effect = .elevatorClosure
        alert.header = "Ruggles elevator 321 (Orange Line Platform to lobby) unavailable due to maintenance"
    }

    let now = EasternTimeInstant.now()

    ScrollView {
        VStack {
            AlertListContainer(
                displayAlerts: .init(highPriority: [.init(alert: serviceChange, isDownstream: false)],
                                     lowPriority: [.init(
                                         alert: elevator,
                                         isDownstream: false
                                     )]),
                showNotAccessibleCard: true,
                now: now,
                isAllServiceDisrupted: false,
                routeIdMatcher: MatcherWildcard(),
                stopIdMatcher: MatcherWildcard(),
                directionIdMatcher: MatcherWildcard(),
                tripId: nil,
                routeAccents: .init(),
                onRowTap: { _, _ in }
            )

            AlertListContainer(
                displayAlerts: .init(highPriority: [.init(alert: elevator, isDownstream: false)], lowPriority: []),
                showNotAccessibleCard: false,
                now: now,
                isAllServiceDisrupted: false,
                routeIdMatcher: MatcherWildcard(),
                stopIdMatcher: MatcherWildcard(),
                directionIdMatcher: MatcherWildcard(),
                tripId: nil,
                routeAccents: .init(),
                onRowTap: { _, _ in }
            )

            AlertListContainer(
                displayAlerts: .init(highPriority: [], lowPriority: [.init(alert: elevator, isDownstream: false)]),
                showNotAccessibleCard: false,
                now: now,
                isAllServiceDisrupted: false,
                routeIdMatcher: MatcherWildcard(),
                stopIdMatcher: MatcherWildcard(),
                directionIdMatcher: MatcherWildcard(),
                tripId: nil,
                routeAccents: .init(),
                onRowTap: { _, _ in }
            )
        }
    }
    .background(Color.blue)
}
